### PR TITLE
ou_getModules: Should fetch a pageable API instead of 100 courses max.

### PR DESCRIPTION
**Pull request by mpellicer**

### CODE REVIEW

File: canvas-where-am-I.js [hunk 1]
```diff
@@ -131,8 +131,48 @@
...
+          'Accept': 'application/json',
+        }
+      }).then(ou_status)
+      .then(async response => {
+        // As the API is pageable, we need to get the modules for each page.
+        // Gets the link header.
+        const linkHeader = response.headers.get('link');
+        // As we do not use external libraries, we get the last page manually.
+        const parts = linkHeader.split(',');
+        const lastPart = parts.find(part => part.includes('rel="last"'));
```
Comment (buckett) [inline canvas-where-am-I.js:149]: This isn't ideal as in https://canvas.instructure.com/doc/api/file.pagination.html they say we shouldn't assume anything about the format of the link URLs, and sometimes they may not include the last relation.
Much better to just look at the next link and walk that.

File: canvas-where-am-I.js [hunk 2]
```diff
@@ -131,8 +131,48 @@
      * https://canvas.instructure.com/doc/api/modules.html#Module
      */
     async function ou_getModules(courseId) {
-      // Added &per_page=100, otherwise only returns the first 10
-      const moduleRequest = `/api/v1/courses/${courseId}/modules?include=items&per_page=100`;
+      // The module API is pageable so it needs to be iterative.
+      const moduleRequest = `/api/v1/courses/${courseId}/modules?include=items`;
```
Comment (buckett) [inline canvas-where-am-I.js:135]: For performance when not in development it's worth still having 100 per page unless the frontend is going to progressively display the extra information.

One request for 100 is normally faster than 10 requests for 10.

**Review by buckett — COMMENTED** [review]
The assignments API was recently changed to make the paging IDs no longer a number, so think it's good to be safe and defensively program here.

File: canvas-where-am-I.js (hunk 2 above)
Comment (mpellicer) [reply canvas-where-am-I.js:135]: I thought the same, just left the default pagination size, bad idea.

File: canvas-where-am-I.js (hunk 1 above)
Comment (mpellicer) [reply canvas-where-am-I.js:149]: Ok, I'll work on a better logic to iterate over pages.

File: canvas-where-am-I.js (hunk 2 above)
Comment (mpellicer) [reply canvas-where-am-I.js]: What do you think about this new implementation?

File: canvas-where-am-I.js (hunk 1 above)
Comment (mpellicer) [reply canvas-where-am-I.js]: What do you think about this new implementation?

**Comment by mpellicer** [comment]
Added a new unit test to ensure the API is pageable and has 'next' links.

File: canvas-where-am-I.js [hunk 3]
```diff
@@ -130,18 +130,34 @@
...
         method: 'GET',
         credentials: 'include',
         headers: {
           'Accept': 'application/json',
         }
       }).then(ou_status)
-      .then(ou_json)
-      .then((moduleArray) => {return moduleArray;})
+      .then(async response => {
+        let modulePage = await Promise.resolve(response.json());
```
Comment (buckett) [inline canvas-where-am-I.js:144]: Do you need the `Promise.resolve`? Doesn't `response.json()` return a promise and doesn't `await` wait for it to resolve?

File: canvas-where-am-I.js [hunk 4]
```diff
@@ -130,18 +130,34 @@
...
+        const linkHeader = response.headers.get('link');
+        // As we do not use external libraries, we get the next page manually.
+        const parts = linkHeader.split(',');
+        const nextPage = parts.find(part => part.includes('rel="next"'));
+        if (nextPage) {
+          // Remove the control characters to get the URL.
+          const nextPageUrl = nextPage.replace('>; rel="next"', '').replace('<','');
+          // Get the modules of the next page recursively and add them to the module list.
+          const nextPageModules = await ou_getModules(courseId, nextPageUrl);
+          nextPageModules.forEach(module => modulePage.push(module));
```
Comment (buckett) [inline canvas-where-am-I.js:156]: What about: `modulePage = modulePage.concat(nextPageModules);`?

**Review by buckett — APPROVED** [review]
Nice, thanks.

File: canvas-where-am-I.js (hunk 3 above)
Comment (mpellicer) [reply canvas-where-am-I.js:144]: True.

File: canvas-where-am-I.js (hunk 4 above)
Comment (mpellicer) [reply canvas-where-am-I.js:156]: Damn, I still need to work on my JS skills, thanks!

**Pull request merged**